amm/drink-tests/src/stable_swap_tests.rs

```diff
@@ -0,0 +1,219 @@
+        stable_swap,
+        BOB,
+        1,
+        initial_reserves.to_vec(),
+        bob(),
+    );
+
+    let (amount_out, fee) = stable_swap::swap_exact_in(
+        session,
+        stable_swap.into(),
+        BOB,
+        token_0,        // in
+        token_1,        // out
+        swap_amount_in, // amount_in
+        0,              // min_token_out
+        bob(),
+    )
+    .result
+    .unwrap()
+    .unwrap();
+
+    // check ruterned amount swapped and fee
+    assert_eq!(expected_swap_amount_out, amount_out, "Amount out mismatch");
+    assert_eq!(expected_fee, fee, "Fee mismatch");
+
+    // check if reserves are equal the actual balances
+    let reserves = stable_swap::reserves(session, stable_swap.into());
+    let balance_0 = psp22_utils::balance_of(session, token_0.into(), stable_swap.into());
+    let balance_1 = psp22_utils::balance_of(session, token_1.into(), stable_swap.into());
+    assert_eq!(
+        reserves,
+        vec![balance_0, balance_1],
+        "Balances - reserves mismatch"
+    );
+
+    //check bobs balances
+    let balance_0 = psp22_utils::balance_of(session, token_0.into(), bob());
+    let balance_1 = psp22_utils::balance_of(session, token_1.into(), bob());
+    assert_eq!(
+        [0, expected_swap_amount_out],
+        [balance_0, balance_1],
+        "Incorrect Bob's balances"
+    );
+
+    // check admin fee
+    let admin_fee_lp = psp22_utils::balance_of(session, stable_swap.into(), charlie());
+    let (total_lp_required, lp_fee_part) = stable_swap::remove_liquidity_by_amounts(
+        session,
+        stable_swap.into(),
+        BOB,
+        admin_fee_lp * 2,
+        [0, expected_admin_fee_part].to_vec(),
+        bob(),
+    )
+    .result
+    .unwrap()
+    .unwrap();
+    assert_eq!(total_lp_required - lp_fee_part, admin_fee_lp, "Incorrect admin fee");
+}
+
+// ref https://github.com/ref-finance/ref-contracts/blob/d241d7aeaa6250937b160d56e5c4b5b48d9d97f7/ref-exchange/src/stable_swap/mod.rs#L744
+#[drink::test]
+fn test_stable_swap_exact_in_01(mut session: Session) {
+    setup_test_swap_exact_in(
+        &mut session,
+        [6, 6],                         // decimals
+        [100000000000, 100000000000],   // initial reserves
+        1000,                           // A
+        6,                              // fee BPS
+        2000,                           // admin fee BPS
+        10000000000,                    // swap_amount_in
+        9999495232,                     // expected out (with fee)
+    );
+}
+
+// ref https://github.com/ref-finance/ref-contracts/blob/d241d7aeaa6250937b160d56e5c4b5b48d9d97f7/ref-exchange/src/stable_swap/mod.rs#L763
+#[drink::test]
+fn test_stable_swap_exact_in_02(mut session: Session) {
+    setup_test_swap_exact_in(
+        &mut session,
+        [12, 18],
+        [100000000000000000, 100000000000000000000000],
+        1000,
+        6,
+        2000,
+        10000000000000000,
+        9999495232752197989995,
+    );
+}
+
```

Comment:
Somewhere here we could put in a test with `0` amount being swapped – https://github.com/ref-finance/ref-contracts/blob/d241d7aeaa6250937b160d56e5c4b5b48d9d97f7/ref-exchange/src/stable_swap/mod.rs#L782

Reply:
There are 2 tests in Ref's repo that test with `amount_in=1` (https://github.com/ref-finance/ref-contracts/blob/d241d7aeaa6250937b160d56e5c4b5b48d9d97f7/ref-exchange/src/stable_swap/mod.rs#L820-L854) where it verifies that the `out=0` (due to rounding I suppose) while the 1 token traded is still transferred to the pools' reserves.

Reply:
> Somewhere here we could put in a test with `0` amount being swapped – https://github.com/ref-finance/ref-contracts/blob/d241d7aeaa6250937b160d56e5c4b5b48d9d97f7/ref-exchange/src/stable_swap/mod.rs#L782

Yes we can, however it works slightly differently in our implementation: when `amount_in=0` swap function returns with an error `InsufficientInputAmount`.


> There are 2 tests in Ref's repo that test with `amount_in=1` (https://github.com/ref-finance/ref-contracts/blob/d241d7aeaa6250937b160d56e5c4b5b48d9d97f7/ref-exchange/src/stable_swap/mod.rs#L820-L854) where it verifies that the `out=0` (due to rounding I suppose) while the 1 token traded is still transferred to the pools' reserves.

Yes, we should add similar tests but again, it works slightly differently than in our implementation so the tests should be accordingly adjusted.